### PR TITLE
:sparkles: react/listen dispatch events

### LINKED ISSUE #75: listen hoc should dispatch a "PLUGUED" action
https://github.com/bpetetot/conference-hall/pull/123#issuecomment-365552028

### PR DESCRIPTION
**Pull request by fabienjuif**
fixes #75 

## TODO
 - [x] Tests

### CODE REVIEW

**Comment by bpetetot** [comment]
indeed, we dont know on which component it's added or removed

File: src/react/listen.jsx [hunk 1]
```diff
@@ -11,11 +13,19 @@ export default (listeners = defaultListeners) => WrappedComponent => class exten
   }
 
   componentWillMount() {
-    this.context.store.listeners.add(listeners)
+    const { store } = this.context
+
+    store.dispatch(toAction('ADDING'))
+    store.listeners.add(listeners)
+    store.dispatch(toAction('ADDED'))
```
Comment (bpetetot) [inline src/react/listen.jsx]: And is it really necessary to have `ADDING` and `ADDED` ?

Just having only `@@krm/LISTENERS>ADDED` maybe renamed in `@@krm/LISTENERS>MOUNTED`

File: src/react/listen.jsx [hunk 2]
```diff
@@ -11,11 +13,19 @@ export default (listeners = defaultListeners) => WrappedComponent => class exten
...
   }
 
   componentWillUnmount() {
-    this.context.store.listeners.remove(listeners)
+    const { store } = this.context
+
+    store.dispatch(toAction('REMOVING'))
+    store.listeners.remove(listeners)
+    store.dispatch(toAction('REMOVED'))
```
Comment (bpetetot) [inline src/react/listen.jsx]: And is it really necessary to have `REMOVING` and `REMOVED` ?

Just having only `@@krm/LISTENERS>REMOVED` maybe renamed in `@@krm/LISTENERS>UNMOUNTED`

**Comment by fabienjuif** [comment]
In my mind this was not a problem since you wanted to do this on screen, and only one screen is loaded at a time.

I first wrote an implementation wich can retrieve a name from:
 - A new parameter
 - Fallback to the componnet name

But I didn't like it, at usage:
 - `export default listen(listeners, 'myName')(Component)`
 - or `export default listen(listeners)(Component)`

In the second case the name could be something like : `inject(Jss(form(otherThing(MyComponent))))`

I then think about retrieving the last name, but I can't do that neither.
Maybe you want to use a common component that is wrapped:
```js
const Common = styles(/* */)(shouldOnlyUpdateForPropTypes(Component))
const Wrapped1 = inject(/* */)(Common)
const Wrapped2 = inject(/* */)(Common)

const WithListener1 = listen(listeners)(Wrapped1) // name === Component
const WithListener2 = listen(listeners)(Wrapped2) // name === Component
```

---

Then, if you think this is something to have, what about the first one ?
 - `export default listen(listeners, name)(Component)`
    * `name` as a string
    * `name` as a function, we give props as parameter

File: src/react/listen.jsx (hunk 1 above)
Comment (fabienjuif) [reply src/react/listen.jsx]: Listeners are not `mounted` this is the wrapped component that is mounted.

In first place I wrote:
**mount**
```js
store.listeners.add(listeners)
store.dispatch(toAction('ADDED'))
```

**unmount**
```js
store.dispatch(toAction('REMOVING'))
store.listeners.remove(listeners)
```

So you can catch both events into listeners that are linked.
But I didn't like that one is before, and the other after the linkage.

So I cloned them to have a "standard".

File: src/react/listen.jsx (hunk 2 above)
Comment (fabienjuif) [reply src/react/listen.jsx]: see previous comment

**Comment by bpetetot** [comment]
With that explanation : 
> In my mind this was not a problem since you wanted to do this on screens, and only one screen is loaded at a time.

I'm agree ! In my use cases it will be enough. But I think in some edge cases, it could be necessary to have a name.

The 2 cases seems enough : 
 - `export default listen(listeners, name)(Component)`
    * `name` undefined: `@@krm/LISTENERS>ADDED`
    * `name` as a string:  `@@krm/LISTENERS>ADDED>%%name%%`

**Comment by fabienjuif** [comment]
So we agree and this PR is about the MVP :)
But I can do that in it

**Comment by fabienjuif** [comment]
Are you ok with `REMOVING` and `REMOVED` actions (before/after) ?

**Comment by guillaumecrespel** [comment]
I don't know if it's necessary, just `ADDED` / `REMOVED` it's ok i think

**Comment by fabienjuif** [comment]
REMOVED can't be catch by screen listeners

**Comment by bpetetot** [comment]
IMO, I will only use catchable (by screen listeners and others) actions : `ADDED` and `REMOVING`
I dont see use cases for `ADDING` and `REMOVED`

**Pull request merged**